Detectors/PHOS/reconstruction/src/Clusterer.cxx

```diff
@@ -23,87 +26,140 @@ using namespace o2::phos;
+        o2::ccdb::CcdbApi ccdb;
+        std::map<std::string, std::string> metadata; // do we want to store any meta data?
+        ccdb.init("http://ccdb-test.cern.ch:8080");  // or http://localhost:8080 for a local installation
+        long bcTime = 1;                             //TODO!!! Convert BC time to time o2::InteractionRecord bcTime = digitsTR.front().getBCData() ;
+        mBadMap = ccdb.retrieveFromTFileAny<o2::phos::BadChannelMap>("PHOS/BadMap", metadata, bcTime);
+        mCalibParams = ccdb.retrieveFromTFileAny<o2::phos::CalibParams>("PHOS/Calib", metadata, bcTime);
+        if (!mBadMap) {
+          LOG(FATAL) << "[PHOSCellConverter - run] can not get Bad Map";
+        }
+        if (!mCalibParams) {
+          LOG(FATAL) << "[PHOSCellConverter - run] can not get CalibParams";
+        }
+      }
+    }
 
-  // Unfold overlapped clusters
-  // TODO:
-  // MakeUnfolding();
+    // Collect digits to clusters
+    makeClusters(digits);
 
-  // Calculate properties of collected clusters (Local position, energy, disp etc.)
-  EvalCluProperties(digits, clusters);
+    LOG(DEBUG) << "Number of PHOS clusters" << clusters->size();
+    // Unfold overlapped clusters
+    // Split clusters with several local maxima if necessary
+    if (o2::phos::PHOSSimParams::Instance().mUnfoldClusters) {
+      makeUnfoldings(digits);
+    }
+    // Calculate properties of collected clusters (Local position, energy, disp etc.)
+    evalCluProperties(digits, clusters, dmc, cluMC);
+
+    trigRec->emplace_back(tr.getBCData(), indexStart, clusters->size());
+  }
   LOG(DEBUG) << "PHOS clustrization done";
 }
 //____________________________________________________________________________
-void Clusterer::MakeClusters(const std::vector<Digit>* digits, std::vector<Cluster>* clusters)
+void Clusterer::makeClusters(const std::vector<Digit>* digits)
 {
   // A cluster is defined as a list of neighbour digits
-  const double kClusteringThreshold = 0.050; // TODO: To be read from RecoParam
-  const double kDigitMinEnergy = 0.010;      // TODO: to be implemented as a digit energy cut
 
   // Mark all digits as unused yet
-  const int maxNDigits = 3584; // There is no clusters larger than PHOS module ;)
-  int nDigits = digits->size();
+  const int maxNDigits = 12546; // There is no digits more than in PHOS modules ;)
   bool digitsUsed[maxNDigits];
 
-  for (int i = 0; i < nDigits; i++) {
+  for (int i = 0; i < maxNDigits; i++) {
     digitsUsed[i] = false;
   }
-  int iFirst = 0; // first index of digit which potentially can be a part of cluster
-                  // e.g. first digit in this module, first CPV digit etc.
+  int iFirst = mFirstDigitInEvent; // first index of digit which potentially can be a part of cluster
 
-  for (int i = 0; i < nDigits; i++) {
-    if (digitsUsed[i])
+  for (int i = iFirst; i < mLastDigitInEvent; i++) {
+    if (digitsUsed[i - mFirstDigitInEvent])
       continue;
 
     const Digit* digitSeed = &(digits->at(i));
```

Comment:
... BTW, once you change input vectors to spans, you should query it as a vector rather then a pointer (i.e. simply digits[i]). Also note that ``vector::at(i)`` (as well as a ``span::at(i)`` is slower than the ``vector[i], span[i]`` since the former is doing a range check (not needed in the correct code).